x-pack/legacy/plugins/apm/server/lib/service_map/get_service_map.ts

```diff
@@ -0,0 +1,165 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { PromiseReturnType } from '../../../typings/common';
+import { rangeFilter } from '../helpers/range_filter';
+import {
+  Setup,
+  SetupTimeRange,
+  SetupUIFilters
+} from '../helpers/setup_request';
+import {
+  SERVICE_NAME,
+  SERVICE_ENVIRONMENT,
+  DESTINATION_ADDRESS,
+  CONNECTION_TYPE,
+  CONNECTION_SUBTYPE,
+  CALLEE_NAME,
+  CALLEE_ENVIRONMENT,
+  CONNECTION_UPSTREAM_LIST
+} from '../../../common/elasticsearch_fieldnames';
+
+export interface IEnvOptions {
+  setup: Setup & SetupTimeRange & SetupUIFilters;
+  serviceName?: string;
+  environment?: string;
+}
+
+interface CyNode {
+  id: string;
+}
+
+interface CyEdge {
+  id: string;
+  source: string;
+  target: string;
+}
+
+interface CyElement {
+  data: CyNode | CyEdge;
+}
+
+export type ServiceMapAPIResponse = PromiseReturnType<typeof getServiceMap>;
+export async function getServiceMap({
+  setup,
+  serviceName,
+  environment
+}: IEnvOptions) {
+  const { start, end, client, uiFiltersES, indices } = setup;
+
+  const params = {
+    index: indices.apmServiceConnectionsIndex,
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { range: rangeFilter(start, end) },
+            { exists: { field: CONNECTION_TYPE } },
+            ...uiFiltersES
+          ]
+        }
+      },
+      aggs: {
+        conns: {
+          composite: {
+            sources: [
+              { [SERVICE_NAME]: { terms: { field: SERVICE_NAME } } },
+              {
+                [SERVICE_ENVIRONMENT]: {
+                  terms: { field: SERVICE_ENVIRONMENT, missing_bucket: true }
+                }
+              },
+              {
+                [DESTINATION_ADDRESS]: {
+                  terms: {
+                    field: DESTINATION_ADDRESS
+                  }
+                }
+              },
+              { [CONNECTION_TYPE]: { terms: { field: CONNECTION_TYPE } } }, // will filter out regular spans etc.
+              {
+                [CONNECTION_SUBTYPE]: {
+                  terms: { field: CONNECTION_SUBTYPE, missing_bucket: true }
+                }
+              }
+            ],
+            size: 1000
+          },
+          aggs: {
+            dests: {
+              terms: {
+                field: CALLEE_NAME
+              },
+              aggs: {
+                envs: {
+                  terms: {
+                    field: CALLEE_ENVIRONMENT,
+                    missing: ''
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  if (serviceName || environment) {
+    const upstreamServiceName = serviceName || '*';
+    const upstreamEnvironment = environment || '*';
+
+    params.body.query.bool.filter.push({
+      wildcard: {
+        [CONNECTION_UPSTREAM_LIST]: `${upstreamServiceName}/${upstreamEnvironment}`
```

Comment:
Maybe this discussion has come up before but how much time have we spent looking into having `service.name` and `service.environment` in separate fields?
If the problem is due to querying (because it's an array of objects) could we used nested fields instead?

That would make querying faster and imo simpler (although nested fields does introduce a slightly different syntax)

Reply:
@dgieselaar any opinion on this?

Reply:
perhaps do both? `service.name`, `service.environment`, and `service.identifier` (or something more appropriately named). I'm not sure if we should use nested, aggregations become more complicated and it doesn't work with instance sorting, which we will probably not use in practice but I think the use case for nested needs to be really compelling as there are some drawbacks.

Reply:
My main concern with the current approach is that it is not longer ECS compliant so we can't apply ui filters like normally. Additionally `wildcard` can be terribly slow.

Reply:
I'm fine with having both for possible future use cases. Although currently I don't think we want to query on `CONNECTION_UPSTREAM_LIST`

Reply:
I'll have a proper look now, might be missing some context